### DISCUSSION TITLE
make openwrt  smartdns configuration simple and stupid as configuration on other linux planform

### DISCUSSION THREAD

**Discussion by emacsnt** [pymumu/smartdns#808] [Ideas]
Smartdns is such a powerful and efficient dns tool, but it has some bad image on Openwrt.
I think that there are two reasons cause this result 

1.  it is very hard to understand the behavior and configuration of smartdns luci app.
2.  the usage of iptables make it hard to maintenance and sometimes cause wired behavior.

suggestion:
1. remove smartdns luci app and only provide one configuration file under etc/smartdns like other linux
2. make the /etc/init.d/smartdns simple and only provide start/stop smartdns, remove iptables manipulation.

sample init.d
```
start_smartdns() {
	procd_open_instance
	procd_set_param limits nofile="65535 65535"
	procd_set_param respawn
	procd_set_param command /usr/sbin/smartdns -f -c /etc/smartdns/smartdns.conf
	procd_close_instance
}

start_service() {
	echo "start smartdns"
	start_smartdns
}

stop_service() {
	echo "stop smartdns"
}
```

**Comment by majoson-chen** [comment]
that's a good idea, i thought that if we remove the configure interfere from luci plane, and we provide another program which have web plane to configure and control (including demon and so on) smartdns. this web plane can also provide an easy-use interface on the other platform and system.

**Comment by pymumu** [comment]
What you said should be supported, just don't install uci. 
Everything else is the same as you said

**Reply by emacsnt** [reply]
> What you said should be supported, just don't install uci. 

currently released smartdns installer package for openwrt works closely with luci app.

1, the initial file /etc/init.d/smartdns has been designed to work with /etc/config/smartdns
2. there  are a lot of iptables manipulation in the initial file /etc/init.d/smartdns
3. the configuration file has been split into three parts: custom.conf, address.conf and blacklist-ip.conf but without the main setting which will be dynamically generated by luci app

To be honest,  I used openwrt smartdns one year before, then got confused by it and remove it from my router.  I tried a lot of other open source dns servers, and finally come back to smartdns this month after I read the full configuration file for other linux platform, then I fully understood how smartdns works and appreciated its efficiency and full features.
https://github.com/pymumu/smartdns/blob/master/etc/smartdns/smartdns.conf

Then I spend three days to learn how to package the ipk on openwrt and get my updated version of Makefile, init.d/smartdns and the full configuration file provided to other linux.
```
define Package/smartdns/conffiles
/etc/smartdns/smartdns.conf
endef

define Package/smartdns/install
	$(INSTALL_DIR) $(1)/usr/sbin $(1)/etc/init.d $(1)/etc/smartdns
	$(INSTALL_BIN) $(PKG_BUILD_DIR)/src/smartdns $(1)/usr/sbin/smartdns
	$(INSTALL_BIN) files/etc/init.d/smartdns $(1)/etc/init.d/smartdns
	$(INSTALL_CONF) files/etc/smartdns/smartdns.conf $(1)/etc/smartdns/
endef
```

Above is my personal experience of how I walk to smrtdns step by step, really thanks to @pymumu , share such a powerful tool for us.
